src/ExtractTileOperations.cpp

```diff
@@ -0,0 +1,322 @@
+#include "ExtractTileOperations.h"
+
+#include "IRMatch.h"  // expr_match
+#include "IRMutator.h"
+#include "IROperator.h"  // Expr + Expr
+#include "Util.h"        // ScopedValue
+
+namespace Halide {
+namespace Internal {
+
+using std::string;
+using std::vector;
+
+namespace {
+
+template<int Dim>
+struct Tile {
+    bool result;
+    Expr base;
+    Expr stride[Dim];
+    int extent[Dim];
+};
+
+const auto wild_i32 = Variable::make(Int(32), "*");
+const auto wild_i32x = Variable::make(Int(32, 0), "*");
+
+Tile<2> is_2d_tile_index(const Expr &e) {
+    // ramp(ramp(base, 1, 4), x4(stride), 4)
+    vector<Expr> matches;
+    if (const auto *r1 = e.as<Ramp>()) {
+        if (const auto *r2 = r1->base.as<Ramp>()) {
+            auto ramp_2d_pattern = Ramp::make(Ramp::make(wild_i32, wild_i32, r2->lanes), Broadcast::make(wild_i32, r2->lanes), r1->lanes);
+            if (expr_match(ramp_2d_pattern, e, matches)) {
+                return {true, std::move(matches[0]), {std::move(matches[2]), std::move(matches[1])}, {r1->lanes, r2->lanes}};
+            }
+        }
+    }
+    return {};
+}
+
+Tile<3> is_3d_tile_index(const Expr &e) {
+    vector<Expr> matches;
+    auto add_sub_pattern = (wild_i32x + wild_i32x) - wild_i32x;
+    if (!expr_match(add_sub_pattern, e, matches)) { return {}; }
```

Comment:
Nit: preferred code style in Halide is to avoid single-line if clauses, so instead,

```
    if (!expr_match(add_sub_pattern, e, matches)) { 
        return {}; 
    }
```

Reply:
Done with e3f1ef6